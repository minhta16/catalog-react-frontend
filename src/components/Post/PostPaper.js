import React from 'react';
import PropTypes from 'prop-types';
import { Typography, Divider, Paper, Button } from '@material-ui/core';
import { Link } from 'react-router-dom';

export function PostPaper(props) {
  const { header, description, price, categoryId, categoryName, created } = props;
  return (
    <Paper className="left-margin">
      <Button color="primary" component={Link} to={`/category/${categoryId}`}>
        Back
      </Button>
      <Typography variant="h4">{header}</Typography>
      <Typography variant="subtitle1">{`Created: ${created}`}</Typography>
      <Typography variant="subtitle1">{`Category: ${categoryName}`}</Typography>
      <Divider />
      <Typography variant="body1">{description}</Typography>
      <Typography variant="body1">{`$ ${price / 100}`}</Typography>
    </Paper>
  );
}

PostPaper.propTypes = {
  header: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  categoryId: PropTypes.string.isRequired,
  categoryName: PropTypes.string.isRequired,
  created: PropTypes.string.isRequired,
  price: PropTypes.number.isRequired,
};

export default PostPaper;
