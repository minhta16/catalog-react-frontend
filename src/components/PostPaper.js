import React from 'react';
import PropTypes from 'prop-types';
import { Typography, Divider, Paper, Button } from '@material-ui/core';
import { Link } from 'react-router-dom';

export function PostPaper(props) {
  const { header, body, categoryId } = props;
  return (
    <Paper className="left-margin">
      <Button color="primary" component={Link} to={`/${categoryId}`}>
        Back
      </Button>
      <Typography variant="h4">{header}</Typography>
      <Typography variant="subtitle1">By: Minh Ta</Typography>
      <Divider />
      <Typography variant="body1">{body}</Typography>
    </Paper>
  );
}

PostPaper.propTypes = {
  header: PropTypes.string.isRequired,
  body: PropTypes.string.isRequired,
  categoryId: PropTypes.string.isRequired,
};

export default PostPaper;
