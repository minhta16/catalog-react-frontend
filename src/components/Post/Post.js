import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import PostPaper from 'components/Post/PostPaper';
import { Container } from '@material-ui/core';
import { selectPost } from 'selectors/posts';
import { selectCategory } from 'selectors/categories';
import InfoSnackbar from 'components/Shared/InfoSnackbar';

export class Post extends Component {
  state = {
    openSnackbar: false,
  };

  componentDidMount = () => {
    const { location } = this.props;
    if (location.snackbarMess) {
      this.closeSnackbar(false);
    }
  };

  closeSnackbar = (close) => () => {
    this.setState({
      openSnackbar: !close,
    });
  };

  render() {
    const { match, currentPost, category, location } = this.props;

    const { openSnackbar } = this.state;
    return (
      <Container maxWidth="lg">
        <PostPaper
          header={currentPost.name}
          created={currentPost.created}
          body={currentPost.description}
          categoryName={category.name}
          categoryId={match.params.id}
        />

        <InfoSnackbar
          open={openSnackbar}
          onClose={this.closeSnackbar(true)}
          message={location.snackbarMess}
        />
      </Container>
    );
  }
}

Post.propTypes = {
  match: PropTypes.object.isRequired,
  currentPost: PropTypes.object.isRequired,
  category: PropTypes.object.isRequired,
  location: PropTypes.object.isRequired,
};

const mapStateToProps = (state, ownProps) => ({
  currentPost: selectPost(state, ownProps.match.params.postId),
  category: selectCategory(state, ownProps.match.params.id),
});
Post.propTypes = {};

export default connect(mapStateToProps)(Post);
