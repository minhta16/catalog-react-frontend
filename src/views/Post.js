import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import PostPaper from 'components/PostPaper';
import { Container } from '@material-ui/core';

export function Post(props) {
  const { match, selectedCatItems } = props;
  const currentPost = selectedCatItems[match.params.postId];
  return (
    <Container maxWidth="lg">
      <PostPaper
        header={currentPost.name}
        created={currentPost.created}
        body={currentPost.description}
        categoryId={match.params.id}
      />
    </Container>
  );
}

Post.propTypes = {
  match: PropTypes.shape.isRequired,
  selectedCatItems: PropTypes.shape.isRequired,
};

const mapStateToProps = (state) => ({
  selectedCatItems: state.posts,
});
Post.propTypes = {};

export default connect(mapStateToProps)(Post);
