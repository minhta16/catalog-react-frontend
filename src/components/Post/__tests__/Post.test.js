/* eslint-disable no-undef */
import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import { Redirect } from 'react-router-dom';
import { configure, shallow } from 'enzyme';
import { Post } from 'components/Post/Post';
import { CircularProgress } from '@material-ui/core';
import PostPaper from '../PostPaper';

configure({ adapter: new Adapter() });
describe('components/Post', () => {
  let wrapper;
  let props;
  // let warning;

  // const update = () => {
  //   wrapper.update();
  //   warning = wrapper.find(Warning);
  // };

  const setup = () => {
    wrapper = shallow(<Post {...props} />);
  };

  beforeEach(() => {
    props = {
      match: {
        params: {
          postId: '1',
          id: '1',
        },
      },
      location: {
        snackbarMess: 'tru',
      },
      selectedCatItems: {
        1: {
          name: 'name',
          created: '2019',
          description: 'desc',
          id: '1',
        },
      },
      currentPost: {
        name: 'name',
        created: '2019',
        description: 'desc',
        id: '1',
      },
      category: {
        name: 'name',
      },
      fetchPosts: jest.fn(),
      openSnackbar: jest.fn(),
    };
  });

  it('should render correctly', () => {
    setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('should redirect if state.redirect', () => {
    setup();
    wrapper.setState({
      redirect: true,
    });
    expect(wrapper.find(Redirect).length).toBe(1);
  });

  it('should not openSnackbar if snackbarMess is not present', () => {
    props.location.snackbarMess = '';
    setup();
    expect(props.openSnackbar).not.toHaveBeenCalled();
  });

  it('should openSnackbar if snackbarMess is present', () => {
    props.location.snackbarMess = 'meomeo';
    setup();
    expect(props.openSnackbar).toHaveBeenCalled();
  });

  it('should display postpaper if category and currentPost is present', () => {
    setup();
    expect(wrapper.find(PostPaper).length).toBe(1);
  });

  it('should not display postpaper if category and currentPost is not present', () => {
    props.currentPost = undefined;
    setup();
    expect(wrapper.find(CircularProgress).length).toBe(1);
  });
});
