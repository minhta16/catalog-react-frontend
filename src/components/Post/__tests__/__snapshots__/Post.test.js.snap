// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Post should render correctly 1`] = `
<WithStyles(ForwardRef(Container))
  maxWidth="lg"
>
  <PostPaper
    categoryId="1"
    categoryName="name"
    created="2019"
    description="desc"
    header="name"
    price={99}
  />
</WithStyles(ForwardRef(Container))>
`;
